cmd/keymasterd/unseal.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+)
+
+func (state *RuntimeState) secretInjectorHandler(w http.ResponseWriter,
+	r *http.Request) {
+	// checks this is only allowed when using TLS client certs.. all other authn
+	// mechanisms are considered invalid... for now no authz mechanisms are in
+	// place i.e. Any user with a valid cert can use this handler
+	if r.TLS == nil {
+		state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+		logger.Printf("We require TLS\n")
+		return
+	}
+	if len(r.TLS.VerifiedChains) < 1 {
+		state.writeFailureResponse(w, r, http.StatusForbidden, "")
+		logger.Printf("Forbidden\n")
+		return
+	}
+	clientName := r.TLS.VerifiedChains[0][0].Subject.CommonName
+	logger.Printf("Got connection from %s", clientName)
+	r.ParseForm()
+	sshCAPassword, ok := r.Form["ssh_ca_password"]
+	if !ok {
+		state.writeFailureResponse(w, r, http.StatusBadRequest,
+			"Invalid Post, missing data")
+		logger.Printf("missing ssh_ca_password")
+		return
+	}
+	if err := state.unsealCA([]byte(sshCAPassword[0]), clientName); err != nil {
+		state.writeFailureResponse(w, r, http.StatusBadRequest,
+			"Invalid Post, "+err.Error())
+		logger.Println(err)
+		return
+	}
+	w.WriteHeader(200)
+	fmt.Fprintf(w, "OK\n")
+	//fmt.Fprintf(w, "%+v\n", r.TLS)
+}
+
+func (state *RuntimeState) tryAutoUnseal() error {
+	baseConfig := state.Config.Base
+	if baseConfig.UnsealPasswordSecretArn == "" {
+		return nil
+	}
+	secretArn, err := arn.Parse(baseConfig.UnsealPasswordSecretArn)
+	if err != nil {
+		return err
+	}
+	if baseConfig.UnsealPasswordSecretKey == "" {
+		baseConfig.UnsealPasswordSecretKey = "UnsealPassword"
+	}
+	creds := credentials.NewCredentials(&ec2rolecreds.EC2RoleProvider{
+		Client:       ec2metadata.New(session.New(&aws.Config{})),
+		ExpiryWindow: time.Minute,
+	})
+	logger.Debugln(0, "getting EC2 role credentials")
+	if value, err := creds.Get(); err != nil {
+		return fmt.Errorf("error getting credentials: %s", err)
+	} else {
+		logger.Debugf(0, "obtained credentials from: %s\n", value.ProviderName)
+	}
+	awsSession, err := session.NewSession(
+		aws.NewConfig().WithCredentials(creds).WithRegion(secretArn.Region))
+	if err != nil {
+		return fmt.Errorf("error creating session: %s", err)
+	}
+	if awsSession == nil {
+		return errors.New("awsSession == nil")
+	}
+	awsService := secretsmanager.New(awsSession)
+	input := secretsmanager.GetSecretValueInput{
+		SecretId: aws.String(baseConfig.UnsealPasswordSecretArn),
+	}
+	output, err := awsService.GetSecretValue(&input)
+	if err != nil {
+		return fmt.Errorf("error calling secretsmanager:GetSecretValue: %s",
+			err)
+	}
+	var secret []byte
+	if output.SecretString != nil {
+		secret = []byte(*output.SecretString)
+	} else {
```

Comment:
make it so that we only allow string secrets

Reply:
Done.